src/pathspec.c

```diff
@@ -550,7 +550,7 @@ int git_pathspec_match_index(
 
 	iter_opts.flags = pathspec_match_iter_flags(flags);
 
-	if (!(error = git_iterator_for_index(&iter, index, &iter_opts))) {
+	if (!(error = git_iterator_for_index(&iter, git_index_owner(index), index, &iter_opts))) {
```

Comment:
This again is prone to segfaulting if the index passed to `git_pathspec_match_index` is unowned.